File: server/verify-seed.js
import dotenv from "dotenv";
import mysql from "mysql2/promise";

dotenv.config();

const DB_HOST = process.env.DB_HOST ?? "127.0.0.1";
const DB_PORT = Number.parseInt(process.env.DB_PORT ?? "3306", 10);
const DB_USER = process.env.DB_USER ?? "root";
const DB_PASSWORD = process.env.DB_PASSWORD ?? "";
const DB_NAME = process.env.DB_NAME ?? "ayuda";

const repeatArg = process.argv.find((item) => item.startsWith("--repeat="));
const repeatCount = repeatArg ? Number.parseInt(repeatArg.split("=")[1] ?? "1", 10) : 1;

function assertCondition(condition, message) {
  if (!condition) {
    throw new Error(message);
  }
}

async function runChecks(pool, iteration) {
  const [projectSummaryRows] = await pool.query(
    `
      SELECT
        COUNT(*) AS totalProjects,
        SUM(CASE WHEN publish_state = 'published' THEN 1 ELSE 0 END) AS publishedProjects,
        SUM(CASE WHEN publish_state = 'draft' THEN 1 ELSE 0 END) AS draftProjects,
        SUM(CASE WHEN status = 'upcoming' THEN 1 ELSE 0 END) AS upcomingProjects,
        SUM(CASE WHEN status = 'active' THEN 1 ELSE 0 END) AS activeProjects,
        SUM(CASE WHEN status = 'archived' THEN 1 ELSE 0 END) AS archivedProjects,
        SUM(CASE WHEN schedule_at IS NULL THEN 1 ELSE 0 END) AS unscheduledProjects,
        SUM(CASE WHEN lat IS NOT NULL AND lng IS NOT NULL THEN 1 ELSE 0 END) AS geocodedProjects,
        SUM(CASE WHEN maps_url IS NOT NULL AND maps_url <> '' THEN 1 ELSE 0 END) AS withMapsUrl
      FROM projects
    `,
  );

  const summary = projectSummaryRows[0];

  assertCondition(Number(summary.totalProjects) >= 8, "Expected at least 8 projects.");
  assertCondition(Number(summary.publishedProjects) >= 1, "Expected at least one published project.");
  assertCondition(Number(summary.draftProjects) >= 1, "Expected at least one draft project.");
  assertCondition(Number(summary.upcomingProjects) >= 1, "Expected at least one upcoming project.");
  assertCondition(Number(summary.activeProjects) >= 1, "Expected at least one active project.");
  assertCondition(Number(summary.archivedProjects) >= 1, "Expected at least one archived project.");
  assertCondition(Number(summary.unscheduledProjects) >= 1, "Expected at least one unscheduled draft project.");
  assertCondition(Number(summary.geocodedProjects) >= 1, "Expected at least one geocoded location.");
  assertCondition(Number(summary.withMapsUrl) >= 1, "Expected at least one project with maps URL.");

  const [requirementRows] = await pool.query("SELECT COUNT(*) AS countValue FROM project_requirements");
  const [eligibilityRows] = await pool.query("SELECT COUNT(*) AS countValue FROM project_eligibility");
  const [dependencyRows] = await pool.query("SELECT COUNT(*) AS countValue FROM project_dependencies");

  assertCondition(Number(requirementRows[0].countValue) >= 8, "Expected mock requirements to be populated.");
  assertCondition(Number(eligibilityRows[0].countValue) >= 8, "Expected mock eligibility rules to be populated.");
  assertCondition(Number(dependencyRows[0].countValue) >= 8, "Expected mock dependencies to be populated.");

  const [readyStateRows] = await pool.query(
    `
      SELECT
        SUM(CASE WHEN ready_summary.all_ready = 1 THEN 1 ELSE 0 END) AS readyProjects,
        SUM(CASE WHEN ready_summary.all_ready = 0 THEN 1 ELSE 0 END) AS blockedProjects
      FROM (
        SELECT p.id, MIN(COALESCE(d.ready, 0)) AS all_ready
        FROM projects p
        LEFT JOIN project_dependencies d ON d.project_id = p.id
        GROUP BY p.id
      ) ready_summary
    `,
  );

  const readiness = readyStateRows[0];
  assertCondition(Number(readiness.readyProjects) >= 1, "Expected at least one ready project.");
  assertCondition(Number(readiness.blockedProjects) >= 1, "Expected at least one blocked project.");

  const [publishedReadyRows] = await pool.query(
    `
      SELECT COUNT(*) AS countValue
      FROM projects p
      WHERE p.publish_state = 'published'
        AND EXISTS (
          SELECT 1
          FROM project_dependencies d
          WHERE d.project_id = p.id
        )
    `,
  );

  assertCondition(
    Number(publishedReadyRows[0].countValue) >= 1,
    "Expected published projects to have dependency checklist values.",
  );

  console.log(
    `Iteration ${iteration}: OK (projects=${summary.totalProjects}, published=${summary.publishedProjects}, draft=${summary.draftProjects})`,
  );
}

async function main() {
  if (!Number.isFinite(repeatCount) || repeatCount < 1) {
    throw new Error("Repeat count must be a positive integer.");
  }

  const pool = mysql.createPool({
    host: DB_HOST,
    port: DB_PORT,
    user: DB_USER,
    password: DB_PASSWORD,
    database: DB_NAME,
    waitForConnections: true,
    connectionLimit: 8,
    queueLimit: 0,
  });

  try {
    for (let iteration = 1; iteration <= repeatCount; iteration += 1) {
      await runChecks(pool, iteration);
    }

    console.log(`All checks passed ${repeatCount} time(s).`);
  } finally {
    await pool.end();
  }
}

main().catch((error) => {
  console.error("Database verification failed.");
  console.error(error.message);
  process.exit(1);
});

File: backend/server/verify_seed.py
from __future__ import annotations

import argparse
from typing import Any

from .db import get_connection


def _as_number(value: Any) -> int:
    return int(value or 0)


def assert_condition(condition: bool, message: str) -> None:
    if not condition:
        raise RuntimeError(message)


def run_checks(cursor, iteration: int) -> None:
    cursor.execute(
        """
        SELECT
          COUNT(*) AS totalProjects,
          SUM(CASE WHEN publish_state = 'published' THEN 1 ELSE 0 END) AS publishedProjects,
          SUM(CASE WHEN publish_state = 'draft' THEN 1 ELSE 0 END) AS draftProjects,
          SUM(CASE WHEN publish_state NOT IN ('draft', 'published') OR publish_state IS NULL OR publish_state = '' THEN 1 ELSE 0 END) AS invalidPublishStateProjects,
          SUM(CASE WHEN status = 'upcoming' THEN 1 ELSE 0 END) AS upcomingProjects,
          SUM(CASE WHEN status = 'active' THEN 1 ELSE 0 END) AS activeProjects,
          SUM(CASE WHEN status = 'archived' THEN 1 ELSE 0 END) AS archivedProjects,
          SUM(CASE WHEN status NOT IN ('upcoming', 'active', 'archived') OR status IS NULL OR status = '' THEN 1 ELSE 0 END) AS invalidStatusProjects,
          SUM(CASE WHEN schedule_at IS NULL THEN 1 ELSE 0 END) AS unscheduledProjects,
          SUM(CASE WHEN lat IS NOT NULL AND lng IS NOT NULL THEN 1 ELSE 0 END) AS geocodedProjects,
          SUM(CASE WHEN maps_url IS NOT NULL AND maps_url <> '' THEN 1 ELSE 0 END) AS withMapsUrl
        FROM projects
        """
    )
    summary = cursor.fetchone()

    assert_condition(_as_number(summary["totalProjects"]) >= 8, "Expected at least 8 projects.")
    assert_condition(_as_number(summary["publishedProjects"]) >= 1, "Expected at least one published project.")
    assert_condition(_as_number(summary["draftProjects"]) >= 1, "Expected at least one draft project.")
    assert_condition(_as_number(summary["invalidPublishStateProjects"]) == 0, "Expected no invalid publish_state values.")
    assert_condition(_as_number(summary["upcomingProjects"]) >= 1, "Expected at least one upcoming project.")
    assert_condition(_as_number(summary["activeProjects"]) >= 1, "Expected at least one active project.")
    assert_condition(_as_number(summary["archivedProjects"]) >= 1, "Expected at least one archived project.")
    assert_condition(_as_number(summary["invalidStatusProjects"]) == 0, "Expected no invalid status values.")
    assert_condition(_as_number(summary["unscheduledProjects"]) >= 1, "Expected at least one unscheduled draft project.")
    assert_condition(_as_number(summary["geocodedProjects"]) >= 1, "Expected at least one geocoded location.")
    assert_condition(_as_number(summary["withMapsUrl"]) >= 1, "Expected at least one project with maps URL.")

    cursor.execute("SELECT COUNT(*) AS countValue FROM project_requirements")
    requirement_count = cursor.fetchone()
    cursor.execute("SELECT COUNT(*) AS countValue FROM project_eligibility")
    eligibility_count = cursor.fetchone()
    cursor.execute("SELECT COUNT(*) AS countValue FROM project_dependencies")
    dependency_count = cursor.fetchone()

    assert_condition(_as_number(requirement_count["countValue"]) >= 8, "Expected mock requirements to be populated.")
    assert_condition(_as_number(eligibility_count["countValue"]) >= 8, "Expected mock eligibility rules to be populated.")
    assert_condition(_as_number(dependency_count["countValue"]) >= 8, "Expected mock dependencies to be populated.")

    cursor.execute(
        """
        SELECT
          SUM(CASE WHEN ready_summary.all_ready = 1 THEN 1 ELSE 0 END) AS readyProjects,
          SUM(CASE WHEN ready_summary.all_ready = 0 THEN 1 ELSE 0 END) AS blockedProjects
        FROM (
          SELECT p.id, MIN(COALESCE(d.ready, 0)) AS all_ready
          FROM projects p
          LEFT JOIN project_dependencies d ON d.project_id = p.id
          GROUP BY p.id
        ) ready_summary
        """
    )
    readiness = cursor.fetchone()
    assert_condition(_as_number(readiness["readyProjects"]) >= 1, "Expected at least one ready project.")
    assert_condition(_as_number(readiness["blockedProjects"]) >= 1, "Expected at least one blocked project.")

    cursor.execute(
        """
        SELECT COUNT(*) AS countValue
        FROM projects p
        WHERE p.publish_state = 'published'
          AND EXISTS (
            SELECT 1
            FROM project_dependencies d
            WHERE d.project_id = p.id
          )
        """
    )
    published_ready = cursor.fetchone()
    assert_condition(
        _as_number(published_ready["countValue"]) >= 1,
        "Expected published projects to have dependency checklist values.",
    )

    print(
        "Iteration "
        f"{iteration}: OK (projects={summary['totalProjects']}, "
        f"published={summary['publishedProjects']}, draft={summary['draftProjects']})"
    )


def main() -> None:
    parser = argparse.ArgumentParser(description="Verify seeded Ayuda database data.")
    parser.add_argument("--repeat", type=int, default=1)
    args = parser.parse_args()

    if args.repeat < 1:
        raise RuntimeError("Repeat count must be a positive integer.")

    connection = get_connection(dict_cursor=True)

    try:
        with connection.cursor() as cursor:
            for iteration in range(1, args.repeat + 1):
                run_checks(cursor, iteration)

        print(f"All checks passed {args.repeat} time(s).")
    finally:
        connection.close()


if __name__ == "__main__":
    try:
        main()
    except Exception as error:
        print("Database verification failed.")
        print(error)
        raise SystemExit(1) from error
